stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,720 @@
+                    &block_receipt.tx_receipts,
+                    &stacks_epoch.block_limit,
+                    &stacks_epoch.epoch_id,
+                );
+            }
+
+            // update fee estimator
+            if let Some(ref mut estimator) = self.fee_estimator {
+                let stacks_epoch = self
+                    .sortition_db
+                    .index_conn()
+                    .get_stacks_epoch_by_epoch_id(&block_receipt.evaluated_epoch)
+                    .expect("Could not find a stacks epoch.");
+                if let Err(e) = estimator.notify_block(&block_receipt, &stacks_epoch.block_limit) {
+                    warn!("FeeEstimator failed to process block receipt";
+                          "stacks_block" => %block_hash,
+                          "stacks_height" => %block_receipt.header.stacks_block_height,
+                          "error" => %e);
+                }
+            }
+
+            let stacks_sn = SortitionDB::get_block_snapshot_consensus(
+                &self.sortition_db.conn(),
+                &canonical_stacks_consensus_hash,
+            )?
+            .expect(&format!(
+                "FATAL: unreachable: consensus hash {} has no snapshot",
+                &canonical_stacks_consensus_hash
+            ));
+
+            // are we in the prepare phase?
+            if !self.burnchain.is_in_prepare_phase(stacks_sn.block_height) {
+                // next ready stacks block
+                continue;
+            }
+
+            // is the upcoming reward cycle processed yet?
+            let current_reward_cycle = self
+                .burnchain
+                .block_height_to_reward_cycle(stacks_sn.block_height)
+                .expect(&format!(
+                    "FATAL: unreachable: burnchain block height has no reward cycle"
+                ));
+
+            let last_processed_reward_cycle = {
+                let ic = self.sortition_db.index_handle(&canonical_sortition_tip);
+                ic.get_last_processed_reward_cycle()?
+            };
+
+            if last_processed_reward_cycle > current_reward_cycle {
+                // already processed upcoming reward cycle
+                continue;
+            }
+
+            // This is the first Stacks block in the prepare phase for the next reward cycle.
+            // Pause here and process the next sortitions
+            return Ok(Some(block_hash));
+        }
+
+        // no PoX anchor block found
+        Ok(None)
+    }
+
+    /// Given a burnchain header, find the PoX reward cycle info
+    pub fn get_nakamoto_reward_cycle_info(
+        &mut self,
+        burn_header: &BurnchainBlockHeader,
+    ) -> Result<Option<RewardCycleInfo>, Error> {
+        let sortition_tip_id = self
+            .canonical_sortition_tip
+            .as_ref()
+            .expect("FATAL: Processing anchor block, but no known sortition tip");
+
+        get_nakamoto_reward_cycle_info(
```

Comment:
Seems a bit strange to keep this separated into two functions like this with the same name? Is this to accommodate easier testing?

Reply:
At one point, I had `get_nakamoto_next_recipients` call it (which doesn't have a handle to the coordinator instance).  I'll just move the code body back.